### PR TITLE
Add support for Markdown Tables

### PR DESCRIPTION
**Pull request by christiankm**
After lots of trial and error (getting to know how the reader works), I have finally succeeded in making a basic implementation for Markdown Tables.

This does at this stage not support all formats out there. It support the general table format which is the one I personally use.

Tables can be with or without headers.

Improvements and suggestions are most welcome!

### CODE REVIEW

**Comment by christiankm** [comment]
@JohnSundell I shall need your help with this implementation. Currently all tests passes and it works fine if the file only contains the table.

However, if it is surrounded by text, the output becomes: 
`<p><table>...</table></p>`

I don't suspect that's what we want, is it? When inside a <p>, the table gets formatted as clear text.

File: Sources/Ink/Internal/Table.swift [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                let columns = line.split(separator: "|")
+                code += "<tr>"
+                columns.forEach { cell in
+                    let trimmedCell = cell.trimmingLeadingWhitespaces().trimmingTrailingWhitespaces()
+                    code += "<td>\(trimmedCell)</td>"
+                }
+
+                code += "</tr>"
+            default:
+                throw Reader.Error()
```
[the remark this answers is not in the document]
Comment (christiankm) [reply Sources/Ink/Internal/Table.swift]: This is the best PR comment I have ever gotten — thank you @john-mueller!

I will implement your suggestions as well as additional tests with fresh eyes in the morning 😌

File: Sources/Ink/Internal/Table.swift [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        reader.rewindIndex()
+
+        var code = ""
+
+        outerWhile: while !reader.didReachEnd {
+            switch reader.currentCharacter {
+            case \.isNewline:
+                code += "</tr>"
+                reader.advanceIndex(by: 2)
+                break
```
[the remark this answers is not in the document]
Comment (christiankm) [reply Sources/Ink/Internal/Table.swift]: These lines were actually called, but if they are removed, tests pass as well, so they were redundant. Thanks!

File: Sources/Ink/Internal/Table.swift [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import Foundation
+
+struct Table: Fragment {
+    var modifierTarget: Modifier.Target { .tables }
+
+    private var code: String
+
+    static func read(using reader: inout Reader) throws -> Table {
+        try reader.read("|")
+        reader.rewindIndex()
```
[the remark this answers is not in the document]
Comment (christiankm) [reply Sources/Ink/Internal/Table.swift]: I see, you are totally right! :)

**Comment by christiankm** [comment]
@john-mueller I have updated my PR with your changes, and added tests. Everything seems well now! Thanks you!

For any advice on, how to make the code O(n), please let's talk.

**Comment by john-mueller** [comment]
I'll take a look tonight or tomorrow. I have a rough idea in my head, but I need to sit down and actually try to prototype it in code.

**Comment by john-mueller** [comment]
@christiankm Here's what I would try next.

### Separating Concerns

First, I'd cleanly separate parsing from rendering. Currently, as you read each line, you are converting it directly into HTML. This has a couple problems:

1. It doesn't parse inline elements (e.g. `*emphasis*` won't be rendered as `<em>emphasis</em>`).
2. Even if we did parse inline elements, we couldn't apply modifiers to them in the `.html(usingURLs:modifiers:)` method because they have already been converted to HTML.

So instead of declaring the `private var code: String` property in `Table`, let's let the internal representation of a table be 

``` swift
private var header: [FormattedText]
private var body: [[FormattedText]]
```

In other words, we have an array of `FormattedText` instances, which represents the cells in the header row, and an array of arrays of `FormattedText`, representing lines of cells of the table. If `header` or `body` are empty, this indicates that the table has no header or no body, respectively. If both `head` and `body` are empty, we should `throw Reader.Error()` rather than return a `Table`.

So the Markdown

```
| Table  | Header     | [Link](/uri) |
| ------ | ---------- | ------------ |
| Some   | *emphasis* | and          |
| `code` | in         | table        |
```

would correspond to a `Table` instance where 

```
header = [
    FormattedText("Table"),
    FormattedText("Header"),
    FormattedText("[Link](/uri)")
]

body = [
    [
        FormattedText("Some"),
        FormattedText("*emphasis*"),
        FormattedText("and")
    ],
    [
        FormattedText("`code`"),
        FormattedText("in"),
        FormattedText("table")
    ]
]
```

Note: There is no initializer called `FormattedText(_:)`; this is just a way to represent the `FormattedText` instances.

### Rendering

Once the `Table` struct has been modified, we can change the `.html(usingURLs:modifiers:)` method to transform the `header` and `body` properties into an HTML string. This is probably most naturally done using `map` and `reduce`, but it doesn't have to be. For the example `Table` struct given above, we would expect the following HTML (Ink doesn't typically add newlines between tags, I just did it here for clarity):

```
<table>
  <thead>
    <tr>
      <th>Table</th>
      <th>Header</th>
      <th><a href="/uri">Link</a></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Some</td>
      <td><em>emphasis</em></td>
      <td>and</td>
    </tr>
    <tr>
      <td><code>code</code></td>
      <td>in</td>
      <td>table</td>
    </tr>
  </tbody>
</table>
```

Note that the `FormattedText` instances have had their own `.html(usingURLs:modifiers:)` method called before being wrapped in `<tr>` or `<td>`. I also added the `<thead>` and `<tbody>` tags, as required by the [GitHub Flavored Markdown Spec](https://github.github.com/gfm/#tables-extension-).

### Parsing

Once the rendering works from a `Table` instance, it will be time to work on the parser. The basic idea would that within the `while` loop, you check that the current line begins with a `"|"`, then start reading `FormattedText` sections terminated by either `"|"` or `"\n"`. If the section ends with a `"|"`, you append that `FormattedText` to the last element of `body`. If the section ends with `"\n"`, you advance the reader to the beginning of the next line, and append an empty array to `body`, to begin the next row. Parsing ends when the current character is no longer a `"|"`.

Because we are relying on the reader to parse through each line, we're only reading things once, instead of reading a line, then reading over it again every time we filter, split, etc.

Once the `while` loop exits, you can check whether the second row consists only of cells containing `"-"`, and if so, place the first element of `body` in `head` and discarding the second.

`FormattedText` has the static method `.read(using:terminator:)`. You might be able to work out a way to stop on *either* a pipe or newline character using existing methods, but I suspect it might be easier to modify the `terminator` parameter (and the underlying `FormattedText.Parser`) to use a set of terminators (`Set<Character>`) instead.

Good luck, and let me know if you get stuck or need a second pair of eyes!

**Comment by john-mueller** [comment]
@christiankm I had some time to sit down and actually try updating your implementation with my suggestions above. The result is [this branch](https://github.com/john-mueller/Ink/tree/markdown-tables-fixup). 

If you want to go through the process of trying to implement it yourself (my recommendation if you have the time and want to learn), feel free just to refer to it for hints if you get stuck. Otherwise, feel free to merge it into your PR.

**Comment by christiankm** [comment]
@john-mueller Thanks you so much for your comments and your changes. I will try to implement your branch into my PR.
I noticed the missing support for inline HTML elements as well, like emphasis which I use in my tables on my blog.

**Comment by christiankm** [comment]
@JohnSundell I have merged @john-mueller s changes into my PR. I think we have a really great implementation now.
Waiting for your review and merge :)

**Comment by fishcharlie** [comment]
@JohnSundell Any plans to review/merge this?

**Comment by christiankm** [comment]
> @JohnSundell Any plans to review/merge this?

He said he has been quite busy lately, and that’s totally understandable in this situation.
I’m sure he’ll get to it soon. In the mean time you can just point to this branch in your Package dependencies, to get the table-support :)
I’ve been doing that for the past few months.

**Comment by fishcharlie** [comment]
> He said he has been quite busy lately, and that’s totally understandable in this situation.
> I’m sure he’ll get to it soon. In the mean time you can just point to this branch in your Package dependencies, to get the table-support :)
> I’ve been doing that for the past few months.

@christiankm However, it's important to note, that @JohnSundell could request breaking changes to this pull request. So it's much better to use a final production version if possible. If @JohnSundell requests breaking changes to this pull request, that would completely break my integration. So it's a tradeoff. I personally think stability is really important, but it is a tradeoff that you take when using new features.

My goal here was just to get some level of insight into the status of this PR. I personally really hesitate when pointing dependencies to a specific branch in my applications. I try to use that as an ultra last resort. It's just really bad practice IMO.

**Comment by christiankm** [comment]
> > He said he has been quite busy lately, and that’s totally understandable in this situation.
> > I’m sure he’ll get to it soon. In the mean time you can just point to this branch in your Package dependencies, to get the table-support :)
> > I’ve been doing that for the past few months.
> 
> @christiankm However, it's important to note, that @JohnSundell could request breaking changes to this pull request. So it's much better to use a final production version if possible. If @JohnSundell requests breaking changes to this pull request, that would completely break my integration. So it's a tradeoff. I personally think stability is really important, but it is a tradeoff that you take when using new features.
> 
> My goal here was just to get some level of insight into the status of this PR. I personally really hesitate when pointing dependencies to a specific branch in my applications. I try to use that as an ultra last resort. It's just really bad practice IMO.

That's perfectly valid, and I agree. I would love to get this reviewed very soon, but sadly it's no longer up to me :)

**Comment by JohnSundell** [comment]
Alright, I've *FINALLY* had some time to sit down and thoroughly review this PR, and in doing so, I also ended up changing it quite a bit. I hope that you don't mind, @christiankm 😅

Basically, I ended up organzing the parsing logic into separate methods, to make the code more self-explanatory (I personally found it very hard to understand the ASCII art that was previously there to illustrate the parsing flow). I also removed all duplicate calls to render each cell as plain text, and made it so that the header and alignment parsing is now done in one pass, rather than having to do two.

I left the tests more or less intact, apart from changing the testing data to things like "HeaderA", and "CellB" (makes for easier debugging, IMO).

The only thing that I changed in terms of the actual parsing was how missing columns are padded. To me, it seems like the correct thing to do would be to insert missing cells, rather than outputting an unbalanced table. But I'd love to hear your opinions on this @christiankm and @john-mueller before we merge this in.

Thanks a lot again for this @christiankm, and I'm terribly sorry that you had to wait so long to have this reviewed, and I hope you don't mind that I changed your implementation a bit.

**Comment by christiankm** [comment]
@JohnSundell Looks good to me! I can hardly see what of my original code is left, but I'm happy as long as it passes tests and works on my site.

I agree on filling out with empty cells. I am not sure if markdown/HTML standards state if an uneven table is actually considered invalid. In that case, maybe it should throw some error instead.

The wait is totally understandable and fine. I've been using my branch in production code, but now I will change to the new version of Ink :)

Ship it! 🚢

**Comment by JohnSundell** [comment]
Thanks @christiankm. I really didn't intend on rewriting the code so much, and it's not that the original code was bad, it's just that now it's a bit more inline with the rest of the library - which makes it easier for me to maintain it going forward. I also chose to rewrite it myself rather than asking you for changes, since you've already had to wait for so long. Glad you like it, and I'll give @john-mueller a chance to weigh in on this too, but will merge it and release it tomorrow unless any serious flaws are found (but given that all of your tests pass, it seems good to go 😀)

**Comment by john-mueller** [comment]
I think it all looks good! Thanks for taking the time to fix it up for merging; the world's been a crazy place, and we appreciate your tooling and content.

**Comment by JohnSundell** [comment]
Thanks a lot for all of your work on this, really happy to finally be able to merge it in 😀

**Comment by JohnSundell** [comment]
@christiankm Since I don't use HTML tables myself anywhere on my websites, do you mind updating your project to use `master`, re-generate your site, and let me know if there are any regressions compared to your implementation? I'd just like to battle test this on a production site before making an official release 🙂

**Comment by JohnSundell** [comment]
All the unit tests pass, so we should be good to go. But you know, always good to do multiple kinds of testing before deploying something to a wider audience 🙂

**Comment by christiankm** [comment]
@JohnSundell Just tested it with master. Looks perfect, and identical! 😍

**Comment by JohnSundell** [comment]
Fantastic, thanks a lot! Release coming right up 🚀

**Pull request merged**